{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1372adc8-685b-43f8-9d54-d57659d8e684",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "753c6709-2d32-4eb7-b422-a7971f5298fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3056317d-bf2d-402f-9d41-69714e286496",
   "metadata": {},
   "source": [
    "# Simulating evolution \n",
    "### (sepcifically) Simulating evolution on a trait that resides on a sex chromosome\n",
    "\n",
    "## Process:\n",
    "\n",
    "Classes that we are defining and using:\n",
    "- fruit fly class, each class object will be an individual:\n",
    "    - sex\n",
    "    - individual type: wild/mutated (on hold for now)\n",
    "    - chromosome1\n",
    "    - chromosome2\n",
    "    - expressed_trait\n",
    " - action: defines actions that are possible in the environment one of them being sex\n",
    "     - sex: takes as an input two individuals (male and female) and returns all possible new chromosomal combinations\n",
    " - chromosome: contains the chromosomal code\n",
    " - fruitfly initializer: intializes the individuals that participate in sex\n",
    "\n",
    "Chromosome class:\n",
    "\n",
    "**Entities:**\n",
    "- chromosomes: type --> panda series\n",
    "    - 10 characters long for X \n",
    "    - 5 characters long for Y\n",
    "        - an allele will simplistically be represented by a single character on this array, which is simple codon (a simplification of a 3 neuclotide codon) \n",
    "        - allele dominance will be an `AND` function\n",
    "        - We will keep all the other alleleles / codons in the letter (AGCT) form and only the sex linked allele will be given a binary codon - 0/1 \n",
    "            - this is so that the AND function works \n",
    "    - for now we are hardcoding mutation site\n",
    "    - at the mutation site, a wildtype will have a `0` and a mutant will have codon `1` (to support the AND function we are going to use)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2180da14-092d-45ec-83bf-14cc302a8a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Chromosome:\n",
    "    def __init__(self, ctype, is_mutant = False):\n",
    "        self.ctype = ctype     # this can be X or Y\n",
    "        self.length = 5 if self.ctype == 'Y' else 10\n",
    "        self.code = self.initialize()\n",
    "        self.is_mutant = is_mutant\n",
    "        self.source = 'wildtype' if is_mutant == False else 'mutant'\n",
    "        self.mutation_site = 7 if self.ctype == 'X' else None      # setting the mutation site on the chromosome, hardcoding it for now\n",
    "        self.add_sex_linked_allele()\n",
    "    \n",
    "    def initialize(self):\n",
    "        dna_neuclotides = 'agct'\n",
    "        data = np.random.randint(0, 4, size=self.length)\n",
    "        dna_code = [dna_neuclotides[d] for d in data]\n",
    "        return pd.Series(dna_code)\n",
    "    \n",
    "    def add_sex_linked_allele(self):\n",
    "        if self.ctype == 'Y':\n",
    "            pass\n",
    "        else:\n",
    "            self.code[self.mutation_site] = 0 if self.source == 'wildtype' else 1\n",
    "        \n",
    "    \n",
    "    def print_chromosome(self):\n",
    "        print(f'Ctype: {self.ctype}')\n",
    "        print(f'length: {self.length}')\n",
    "        print(f'source: {self.source}')\n",
    "        print(f'is_mutant: {self.is_mutant}')\n",
    "        print(f'\\nCode:\\n{self.code}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8860fbc0-b14a-47f0-a646-a8e95921dab0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "749d4726-1fb0-45e5-baea-6819aef62370",
   "metadata": {},
   "source": [
    "Fruit fly Class:\n",
    "\n",
    "Inherits the chromosome class (because chromosomes are required to make an individual)\n",
    "\n",
    "Attributes:\n",
    "- sex: user passed and can be `M` or `F`\n",
    "- is_mutant: 1 or 0 pointing to True or False of if the individual is a mutant type or not\n",
    "- c1,c2: the two chromosomes that form the two chromatids that make us the 23rd chromosome of the fruit fly individual\n",
    "    - if the chromosomes are passed in the class definition then those are used to create the indidividual \n",
    "    - if no chromosomes are passed then a new individual is randomly generated using the random dna generator in the chromosome class\n",
    "- expressed_trait_code, expressed_trait: 0--> `white eyes`, and 1--> `red eyes` based on the chromosomes\n",
    "\n",
    "Funtions in the class:\n",
    "- initializer class: initializes the class variables \n",
    "- initialize chromosomes: is called if the user does not supply a chromosome to craft the individual \n",
    "- expressed_trait, trait_decoder: called to check if the individual has white eyes or red eyes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "79cdd5b1-8d06-49ec-876a-67b09ce136ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Fruit_fly(Chromosome):\n",
    "    def __init__(self, sex = 'F', is_mutant = False, c1 = None, c2 = None):\n",
    "\n",
    "        # Case 1: if no chromatids are passed then GENERATE NEW CHROMATIDS to spawn a NEW individual\n",
    "        if c1 == None and c2 == None:\n",
    "            self.sex = sex\n",
    "            self.is_mutant = is_mutant            \n",
    "            self.c1, self.c2 = self.initialize_chromosomes()\n",
    "            self.expressed_trait_code = self.express_trait()\n",
    "        \n",
    "        # Case 2: If chromatids are passed, then JOIN THE CHROMATIDS to create a new indvidual \n",
    "        else:\n",
    "            self.c1 = c1\n",
    "            self.c2 = c2\n",
    "            self.sex = self.sex_checker()\n",
    "            self.expressed_trait_code = self.express_trait()\n",
    "            self.is_mutant = bool(self.expressed_trait_code)\n",
    "            \n",
    "\n",
    "        self.expressed_trait = self.trait_decoder()\n",
    "\n",
    "        \n",
    "    def initialize_chromosomes(self):\n",
    "        \"\"\"\n",
    "        initialize the two chromatids that form an organism \n",
    "        using the parent class\n",
    "        \"\"\"\n",
    "        ctid1 = Chromosome('X', is_mutant= self.is_mutant)\n",
    "        if self.sex == 'F':\n",
    "            ctid2 = Chromosome('X', is_mutant= self.is_mutant)\n",
    "        else:\n",
    "            ctid2 = Chromosome('Y', is_mutant= self.is_mutant)\n",
    "        return ctid1, ctid2\n",
    "    \n",
    "    \n",
    "    def sex_checker(self):\n",
    "        if self.c1.ctype == self.c2.ctype:\n",
    "            return 'F'\n",
    "        else:\n",
    "            return 'M'\n",
    "        \n",
    "    \n",
    "    def express_trait(self):\n",
    "        if self.sex == 'F':\n",
    "            return self.c1.code[self.c1.mutation_site] & self.c2.code[self.c2.mutation_site]\n",
    "        \n",
    "        # because only one of the chromatids will have the mutation\n",
    "        else:\n",
    "            try:\n",
    "                return self.c1.code[self.c1.mutation_site]\n",
    "            except:\n",
    "                return self.c2.code[self.c2.mutation_site]\n",
    "    \n",
    "    def trait_decoder(self):\n",
    "        return 'white eyes' if self.expressed_trait_code == 1 else 'red eyes'\n",
    "    \n",
    "    def print_properties(self):\n",
    "        print(f'sex: {self.sex}')\n",
    "        print(f'is mutant? {self.is_mutant}')\n",
    "        print(f\"expressed_trait: [{self.expressed_trait}]\")\n",
    "        self.c1.print_chromosome()\n",
    "        self.c2.print_chromosome()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "42f1b0ea-8ecd-44cc-a7c6-5f007e67f4f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Sex(Fruit_fly):\n",
    "    def __init__(self, ind1, ind2):\n",
    "        self.ind1 = ind1\n",
    "        self.ind2 = ind2\n",
    "        \n",
    "        # if they are of the opposite sex, then let them create new individuals\n",
    "        if self.gender_checker():\n",
    "            self.offspring = list(self.sexual_genesis())\n",
    "            # self.off0, self.off1, self.off2, self.off3 = self.sexual_genesis()\n",
    "        else:\n",
    "            print('Sexually Incompatible')\n",
    "    \n",
    "    def sexual_genesis(self):\n",
    "        return Fruit_fly(c1=self.ind1.c1, c2=self.ind2.c1), Fruit_fly(c1=self.ind1.c1, c2=self.ind2.c2), \\\n",
    "        Fruit_fly(c1=self.ind1.c2, c2=self.ind2.c1), Fruit_fly(c1=self.ind1.c2, c2=self.ind2.c2)\n",
    "        # create 4 new individuals\n",
    "        \n",
    "    \n",
    "    def gender_checker(self):\n",
    "        # check that genders are opposite \n",
    "        if self.ind1.sex == self.ind2.sex:\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "    \n",
    "    def logfile(self):\n",
    "        for (i,off) in enumerate(self.offspring):\n",
    "            print(f\"baby #{i}, sex: {str(off.sex)}, trait: {off.expressed_trait}\")\n",
    "            #print(f\"baby #{(i)}\")\n",
    "            #off.print_properties()\n",
    "    \n",
    "        # number of offspring\n",
    "        # capture number of males and females\n",
    "        # number of mutants vs non-mutants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7520abd8-d909-49da-b14b-6cf3fafa8251",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------------------------------------- Starting Simulation ----------------------------------------------------\n",
      "\n",
      "\n",
      "====== starting conditions ======\n",
      "\n",
      "\n",
      "\n",
      "**************** end of current generation run ****************\n",
      "\n",
      "\n",
      "**************** end of current generation run ****************\n",
      "\n",
      "\n",
      "**************** end of current generation run ****************\n",
      "\n",
      "\n",
      "**************** end of current generation run ****************\n"
     ]
    }
   ],
   "source": [
    "# this is where the generations will be created and sex will happen\n",
    "# we will set up each generation into two buckets Male and Female\n",
    "# the zeroth generations will have individuals created / spawned \n",
    "# the subsequent generations will be born from the sex of the male and female species\n",
    "\n",
    "class Environment(Sex):\n",
    "    def __init__(self, num_generations):\n",
    "        \n",
    "        # TODO: Add an assert statement here to check the num_generations > 1\n",
    "        self.num_generations = num_generations\n",
    "        self.generations = self.create_generation()\n",
    "        self.sexually_active_females, self.sexually_active_males = self.spawn_env()\n",
    "        self.offspring_females, self.offspring_males = [],[]\n",
    "        self.start_simulation()\n",
    "\n",
    "        \n",
    "    \n",
    "    def start_simulation(self):\n",
    "        \n",
    "        print('---------------------------------------------------- Starting Simulation ----------------------------------------------------\\n\\n')\n",
    "        print('====== starting conditions ======\\n')\n",
    "        #self.print_env()\n",
    "\n",
    "        while len(self.generations) < self.num_generations:\n",
    "            self.produce_next_gen()\n",
    "            #self.print_env()\n",
    "            print('\\n\\n**************** end of current generation run ****************')\n",
    "        \n",
    "        print('\\n---------------------------------------------------- End of Simulation ----------------------------------------------------')\n",
    "            \n",
    "    \n",
    "    \n",
    "    def create_generation(self):\n",
    "        return []\n",
    "    \n",
    "    def spawn_env(self, mutant_in_mix = True):\n",
    "        # we create the zeroth or the F0 generation as they have to be made\n",
    "        # the first set will be females\n",
    "        females, males = [],[]\n",
    "        females.append(Fruit_fly(sex='F', is_mutant=False))\n",
    "        males.append(Fruit_fly(sex='M', is_mutant=mutant_in_mix))\n",
    "        \n",
    "        return females, males\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "    # Todo: make changes to the print file to account for the sex segregation you are doing \n",
    "    def print_env(self):\n",
    "        # print(self.generations)\n",
    "        for gen in self.generations:\n",
    "            print('\\n+++++++++++++ gen +++++++++++++')\n",
    "            for sexes in gen:\n",
    "                for individuals in sexes:\n",
    "                    print((individuals.sex, individuals.expressed_trait_code), end = '\\t')\n",
    "\n",
    "    \n",
    "    ###### Action functions ######\n",
    "    \n",
    "    # produces the next generation by creating offspring individuals from the sexually active adults\n",
    "    # and completes the generation run by moving offspring to the sexually active list\n",
    "    def produce_next_gen(self): \n",
    "        for f in self.sexually_active_females:\n",
    "            for m in self.sexually_active_males:\n",
    "                s = Sex(f,m)\n",
    "                self.sex_segregator(s.offspring)\n",
    "        \n",
    "        self.complete_current_gen()    \n",
    "    \n",
    "    # at the complete of a gen run we move the offspring to sexually active\n",
    "    # we move sexually active to gen \n",
    "    # (as they are no longer sexually active, they are replaced by their offspring)\n",
    "    # we empty the offspring list, ready for the next birthing cycle\n",
    "    def complete_current_gen(self):\n",
    "        self.generations.append([self.sexually_active_females, self.sexually_active_males])\n",
    "        self.sexually_active_females, self.sexually_active_males = self.offspring_females, self.offspring_males\n",
    "        self.offspring_females, self.offspring_males = [],[]\n",
    "        \n",
    "    \n",
    "    # will file the males and females into separate lists \n",
    "    def sex_segregator(self, offspring_list):\n",
    "        for off in offspring_list:\n",
    "            if off.sex == 'M':\n",
    "                self.offspring_males.append(off)\n",
    "            else:\n",
    "                self.offspring_females.append(off)\n",
    "    \n",
    "    \n",
    "e = Environment(5)      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dbdfda7-eafc-4e42-acf2-7d9440311744",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Logger:\n",
    "    def __init__(self, generations):\n",
    "        self.generations = generations\n",
    "        self.aggregate_metrics_calculator()\n",
    "    \n",
    "    def aggregate_metrics_calculator(self):\n",
    "        for gen in self.generations:\n",
    "            print(f\"F: {len(gen[0])} \\t\\tM: {len(gen[1])}\")\n",
    "            mutant_count, non_mutants = 0,0\n",
    "\n",
    "            for female in gen[0]:\n",
    "                if female.expressed_trait_code == 1:\n",
    "                    mutant_count+=1\n",
    "                else:\n",
    "                    non_mutants+=1\n",
    "            #print(f\"mutants: {mutant_count} non_mutants: {non_mutants}\")\n",
    "            \n",
    "            #mutant_count, non_mutants = self.count_resetter(mutant_count, non_mutants)\n",
    "            \n",
    "            for male in gen[1]:\n",
    "                if male.expressed_trait_code == 1:\n",
    "                    mutant_count+=1\n",
    "                else: \n",
    "                    non_mutants+=1\n",
    "            print(f\"mutants: {mutant_count} non_mutants: {non_mutants}\")\n",
    "    \n",
    "    \n",
    "    def count_resetter(self, a, b):\n",
    "        return 0,0\n",
    "                \n",
    "l = Logger(e.generations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bf43ddec-a82e-4588-9ad7-b49f5c2a6da6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fem_mutants: 12.500000%\n",
      "fem_non_mutants: 87.500000%\n",
      "mal_mutants: 25.000000%\n",
      "mal_non_mutants: 75.000000%\n"
     ]
    }
   ],
   "source": [
    "print(\"fem_mutants: {:%}\".format(16/128))\n",
    "print(\"fem_non_mutants: {:%}\".format(112/128))\n",
    "print(\"mal_mutants: {:%}\".format(32/128))\n",
    "print(\"mal_non_mutants: {:%}\".format(96/128))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3445110a-6c11-4efc-9873-ce7fdd8ea9a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "256"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pow(16,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b6386fd-e68f-438f-a9e9-d94cdb7ef50a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
